projects/transformers/run_utils.py

```diff
@@ -459,6 +484,286 @@ def preprocess_function(examples):
+            if not data_args.beam_search:
+                examples[question_column_name] = [q.lstrip() for q in examples[
+                    question_column_name]]
+
+            tokenized_examples["example_id"] = []
+
+            if data_args.beam_search:
+                special_tokens = tokenized_examples.pop("special_tokens_mask")
+                tokenized_examples["cls_index"] = []
+                tokenized_examples["p_mask"] = []
+
+            context_idx = 1 if pad_on_right else 0
+            for i, input_ids in enumerate(tokenized_examples["input_ids"]):
+
+                if data_args.beam_search:
+                    cls_index = input_ids.index(tokenizer.cls_token_id)
+                    tokenized_examples["cls_index"].append(cls_index)
+                    sequence_ids = tokenized_examples["token_type_ids"][i]
+                    for k, s in enumerate(special_tokens[i]):
+                        if s:
+                            sequence_ids[k] = 3
+                    tokenized_examples["p_mask"].append(
+                        [
+                            0.0 if (not special_tokens[i][k] and s == context_idx) or k == cls_index else 1.0  # noqa: E501
+                            for k, s in enumerate(sequence_ids)
+                        ]
+                    )
+                else:
+                    sequence_ids = tokenized_examples.sequence_ids(i)
+
+                sample_index = sample_mapping[i]
+                tokenized_examples["example_id"].append(examples["id"][sample_index])
+
+                tokenized_examples["offset_mapping"][i] = [
+                    (o if sequence_ids[k] == context_idx else None)
+                    for k, o in enumerate(tokenized_examples["offset_mapping"][i])
+                ]
+
+            return tokenized_examples
+
+    return preprocess_function
+
+
+def prepare_squad_dataset(split,
+                          datasets,
+                          data_args,
+                          prepare_features,
+                          column_names):
+
+    dataset, examples = None, None
+    if split not in datasets:
+        raise ValueError(f"--this run requires a {split} dataset")
+
+    examples = datasets[split]
+    split_key = "eval" if split == "validation" else "train"
+    max_samples_key = f"max_{split_key}_samples"
+
+    if getattr(data_args, max_samples_key) is not None:
+        examples = examples.select(range(getattr(data_args, max_samples_key)))
+
+    dataset = examples.map(
+        prepare_features,
+        batched=True,
+        num_proc=data_args.preprocessing_num_workers,
+        remove_columns=column_names,
+        load_from_cache_file=not data_args.overwrite_cache,
+    )
+
+    if getattr(data_args, max_samples_key) is not None:
+        dataset = dataset.select(range(getattr(data_args, max_samples_key)))
+
+    return examples, dataset
+
+
+def get_squad_tokenizer_kwargs(data_args, pad_on_right, max_seq_length):
```

Comment:
Most of these functions are missing docstrings, would be useful to add at least a one-liner